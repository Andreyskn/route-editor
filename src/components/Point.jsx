import React, { Component } from 'react';

export class Point extends Component {

	render() {
		const { text, isDragged, isClone, offsetTop, onMouseOver, startDragging, onRemove } = this.props;

		const containerProps = {
			className: `point${isClone ? ' clone' : ''}${isDragged ? ' dragged' : ''}`,
			style: offsetTop ? { top: offsetTop } : undefined,
			onMouseOver,
		}

		return (
			<div {...containerProps}>
				<button className='point-btn' onClick={onRemove} />
				<div className='point-overlay' onMouseDown={startDragging}>
					<span className='point-text' title={text}>{text}</span>
				</div>
			</div>
		);
	}
}
