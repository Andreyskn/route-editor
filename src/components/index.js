export * from './PointsList';
export * from './Point';