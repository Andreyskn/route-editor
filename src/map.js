export class MapController {
	constructor(containerId) {
		if (!window.ymaps) throw new Error('Missing Map API script');

		this.map = null;
		this.points = {};
		this.polyline = null;
		this.order = [];

		window.ymaps.ready(() => this.init(containerId));
	}

	init = (containerId) => {
		try {
			this.map = new ymaps.Map(containerId, {
				center: [57.638724, 44.083593],
				zoom: 4,
				controls: [],
			});
		} catch {
			throw new Error(`Element with id="${containerId}" not found in DOM`);
		}
	}

	setActionCallback = (callback) => {
		window.ymaps.ready(() => this.map && this.map.events.add('actionend', callback));
	}

	addPlacemark = (id, name, order) => {
		this.order = order;

		const newPoint = new ymaps.Placemark(
			this.map.getCenter(),
			{ balloonContent: name },
			{ draggable: true, preset: 'islands#blueDotIcon' }
		);
		newPoint.events.add('dragend', this.drawLines);

		this.points[id] = newPoint;
		this.map.geoObjects.add(newPoint);

		this.drawLines();
	}

	removePlacemark = (id, order) => {
		this.order = order;

		this.map.geoObjects.remove(this.points[id]);
		delete this.points[id];

		this.drawLines();
	}

	changeOrder = (order) => {
		this.order = order;

		this.drawLines();
	}

	drawLines = () => {
		this.map.geoObjects.remove(this.polyline);

		this.polyline = new ymaps.Polyline(
			this.order.map(id => this.points[id].geometry.getCoordinates()),
			{},
			{ strokeWidth: 2, strokeColor: '#0000FF' }
		);

		this.map.geoObjects.add(this.polyline);
	}

	togglePlacemarkHighlight = (id, highlight) => {
		this.points[id].options.set('preset', highlight ? 'islands#greenDotIcon' : 'islands#blueDotIcon');
	}
}