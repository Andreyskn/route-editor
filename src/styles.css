*
{
	box-sizing: border-box;
}

body
{
	margin: 0;
}

.app
{
	display: flex;
	overflow: hidden;

	width: 100vw;
	height: 100vh;
}

.sidebar
{
	min-width: 230px;
	max-width: 380px;
	padding: 15px 0 0 15px;

	background-color: #466675;

	flex: 1 0 25%;
}

.input
{
	font-size: 16px;

	width: calc(100% - 15px);
	height: 40px;
	margin-bottom: 15px;
	padding: 0 10px;

	border: 1px solid black;
}

.points-list
{
	position: relative;

	overflow: auto;

	height: calc(100% - 55px);
	padding-right: 15px;
}

.point
{
	position: relative;

	display: flex;

	width: 100%;
	height: 50px;
	padding-right: 5px;

	user-select: none;

	border: 1px solid #b9e3f6;
	border-radius: 3px;
	background-color: aliceblue;

	align-items: center;
	justify-content: flex-end;
}

.point + .point:not(.clone)
{
	margin-top: 5px;
}

.dragged
{
	opacity: 0;
}

.clone
{
	position: absolute;
	top: 0;
	left: 0;

	width: calc(100% - 15px);

	pointer-events: none;

	opacity: .8;
	box-shadow: 1px 2px 5px 0 rgba(0,0,0,.75);
}

.point-overlay
{
	position: absolute;
	top: 0;
	left: 0;

	display: flex;

	width: 100%;
	height: 100%;
	padding-left: 10px;

	align-items: center;
}

.point-text
{
	font-size: 20px;

	display: block;
	overflow: hidden;

	max-width: calc(100% - 50px);

	white-space: nowrap;
	text-overflow: ellipsis;
}

.point-btn
{
	position: relative;
	z-index: 1;

	width: 40px;
	height: 40px;
	padding: 0;

	cursor: pointer;
	transition: all .25s;

	border: none;
	background-color: rgba(0, 0, 0, .04);

	flex-shrink: 0;
}

.point-btn:before,
.point-btn:after
{
	position: absolute;
	top: 50%;
	left: 50%;

	width: 2px;
	height: 15px;

	content: '';
	transition: all .25s;
	transform-origin: 0 0;

	background-color: #999;
}

.point-btn:before
{
	transform: rotate(45deg) translate(-50%, -50%);
}

.point-btn:after
{
	transform: rotate(-45deg) translate(-50%, -50%);
}

.point-btn:hover
{
	background-color: rgba(0, 0, 0, .1);
}

.point-btn:hover:before,
.point-btn:hover:after
{
	background-color: #000;
}

#map
{
	background-color: lightgray;

	flex: 1 0 75%;
}
