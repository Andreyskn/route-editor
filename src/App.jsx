import React, { Component } from 'react';

import { PointsList } from './components';
import MapController from './map';

const ENTER_KEY = 13;

export class App extends Component {

	_mapController = null;
	_input = null;

	state = {
		points: {},
		order: [],
	}

	componentDidMount() {
		this._mapController = new MapController('map');
		this._input && this._mapController.setActionCallback(this.setInputFocus);
	}

	setInputFocus = () => this._input.focus();

	generateId(ids) { return !ids.length ? 0 : Math.max(...ids) + 1 };

	addPoint = (e) => {
		const value = e.target.value;
		if (e.keyCode !== ENTER_KEY || value === '') return;

		const { points, order } = this.state;

		const newId = this.generateId(order);
		const updatedPoints = { ...points, [newId]: value };
		const updatedOrder = [...order, newId];

		e.target.value = '';

		this.setState({ points: updatedPoints, order: updatedOrder });
		this._mapController.addPlacemark(newId, value, updatedOrder);
	}

	removePoint = (id) => () => {
		const { points, order } = this.state;

		delete points[id];
		const updatedOrder = order.filter(i => i !== id);

		this.setState({ points, order: updatedOrder });
		this._mapController.removePlacemark(id, updatedOrder);
	}

	changeOrder = (order) => {
		this.setState({ order });
		this._mapController.changeOrder(order);
	}

	render() {
		const { points, order } = this.state;

		const inputProps = {
			type: 'text', className: 'input',
			placeholder: 'Новая точка маршрута',
			onKeyDown: this.addPoint, autoFocus: true,
		}

		const pointsListProps = {
			points, order,
			changeOrder: this.changeOrder,
			removePoint: this.removePoint,
			mapController: this._mapController,
		}

		return (
			<div className='app'>
				<div className='sidebar'>
					<input {...inputProps} ref={x => this._input = x} />
					<PointsList {...pointsListProps} />
				</div>
				<div id='map' />
			</div>
		)
	}
}