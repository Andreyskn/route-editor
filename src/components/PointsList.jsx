import React, { Component } from 'react';

import { Point } from './Point';

const emptyItem = { id: null, node: null, yPos: 0 };

export class PointsList extends Component {

	_itemInnerOffset = null;
	_scrollContainer = null;
	_itemsContainer = {
		node: null,
		offset: null,
	}

	state = {
		isDragging: false,
		draggedItem: emptyItem,
	}

	startDragging = (id) => (e) => {
		const rect = e.target.getBoundingClientRect();
		this._itemInnerOffset = e.clientY - rect.top;
		this._itemsContainer.offset = this._itemsContainer.node.getBoundingClientRect().top;
		if (!this._scrollContainer) this._scrollContainer = this._itemsContainer.node.parentNode;

		this.setState({
			isDragging: true,
			draggedItem: {
				id,
				node: e.target.parentNode,
				yPos: this.getDragPosition(e.clientY),
			},
		});

		window.addEventListener('mousemove', this.watchMouseMove);
		window.addEventListener('mouseup', this.finishDragging);
		this._scrollContainer.addEventListener('scroll', this.onScroll);
	}

	changeOrder = (targetId) => () => {
		const { order, changeOrder } = this.props;
		const { draggedItem: { id: currentId } } = this.state;

		if (targetId === currentId) return;

		const newOrder = order.map(id => {
			if (id === targetId) return currentId;
			if (id === currentId) return targetId;
			return id;
		})

		changeOrder(newOrder);
	}

	finishDragging = () => {
		this.setState({ isDragging: false, draggedItem: emptyItem });

		window.removeEventListener('mousemove', this.watchMouseMove);
		window.removeEventListener('mouseup', this.finishDragging);
		this._scrollContainer.removeEventListener('scroll', this.onScroll);
	}

	watchMouseMove = (e) => {
		this.setState(s => ({
			draggedItem: {
				...s.draggedItem,
				yPos: this.getDragPosition(e.clientY),
			}
		}));
	}

	onScroll = () => {
		this._itemsContainer.offset = this._itemsContainer.node.getBoundingClientRect().top;
	}

	getDragPosition = (yMousePos) => yMousePos - this._itemInnerOffset - this._itemsContainer.offset;

	toggleItemHover = (id, hover) => () => {
		const { props: { mapController }, state: { isDragging } } = this;

		!isDragging && mapController.togglePlacemarkHighlight(id, hover);
	}

	render() {
		const { points, order, removePoint } = this.props;
		const { isDragging, draggedItem } = this.state;

		const renderPoint = (id) => {
			const pointProps = {
				key: id,
				isDragged: draggedItem.id === id,
				onMouseOver: isDragging ? this.changeOrder(id) : undefined,
				onMouseEnter: this.toggleItemHover(id, true),
				onMouseLeave: this.toggleItemHover(id, false),
				startDragging: this.startDragging(id),
				onRemove: removePoint(id),
				text: points[id],
			}

			return <Point {...pointProps} />;
		}

		return (
			<div className='points-list'>
				<div ref={x => this._itemsContainer.node = x}>
					{order.map(renderPoint)}
					{isDragging && <Point isClone offsetTop={`${draggedItem.yPos}px`} text={draggedItem.node.innerText} />}
				</div>
			</div>
		)
	}
}