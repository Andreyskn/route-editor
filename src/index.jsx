import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import './styles.css';

import { PointsList } from './components';
import { MapController } from './map';

const ENTER_KEY = 13;

class App extends Component {

	_mapController = null;

	state = {
		points: {},
		order: [],
	}

	componentDidMount() {
		this._mapController = new MapController('map');
	}

	generateId = (ids) => !ids.length ? 0 : Math.max(...ids) + 1;

	addPoint = (e) => {
		const value = e.target.value;

		if (e.keyCode === ENTER_KEY && value !== '') {
			const { points, order } = this.state;

			const newId = this.generateId(order);
			const updatedPoints = { ...points, [newId]: value };
			const updatedOrder = [...order, newId];

			e.target.value = '';

			this.setState({ points: updatedPoints, order: updatedOrder });
		}
	}

	removePoint = (id) => () => {
		const { points, order } = this.state;

		delete points[id];
		const updatedOrder = order.filter(i => i !== id);

		this.setState({ points, order: updatedOrder });
	}

	changeOrder = (order) => {
		this.setState({ order });
	}

	render() {
		const { points, order } = this.state;

		return (
			<div className='app'>
				<div className='sidebar'>
					<input type='text' className='input' autoFocus placeholder='Новая точка маршрута' onKeyDown={this.addPoint} />
					<PointsList points={points} order={order} changeOrder={this.changeOrder} removePoint={this.removePoint} />
				</div>
				<div id='map' />
			</div>
		)
	}
}

ReactDOM.render(<App />, document.getElementById('root'));
