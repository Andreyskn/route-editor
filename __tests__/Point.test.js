import React from 'react';
import { shallow } from 'enzyme';
import { Point } from '../src/components';

describe('<Point />', () => {

	const renderPoint = (props = {}) => shallow(<Point {...props}/>);

	describe('Rendering', () => {
		const defaultProps = {
			text: 'Test text',
		}

		it('renders Default state correctly', () => {
			expect(renderPoint(defaultProps)).toMatchSnapshot();
		});

		it('renders Dragged state correctly', () => {
			const dragProps = {
				isDragged: true,
			}

			expect(renderPoint({...defaultProps, ...dragProps})).toMatchSnapshot();
		});

		it('renders Clone state correctly', () => {
			const cloneProps = {
				isClone: true,
				offsetTop: '10px',
			}

			expect(renderPoint({...defaultProps, ...cloneProps})).toMatchSnapshot();
		});
	});

	describe('Mouse handlers', () => {
		const mockFn = jest.fn();

		it('calls onRemove function when remove button is clicked', () => {
			renderPoint({ onRemove: mockFn }).find('.point-btn').simulate('click');
			expect(mockFn).toHaveBeenCalledTimes(1);
		});

		it('calls startDragging function when mousedown event is triggered on point-overlay', () => {
			renderPoint({ startDragging: mockFn }).find('.point-overlay').simulate('mousedown');
			expect(mockFn).toHaveBeenCalledTimes(1);
		});

		it('calls onMouseOver, onMouseEnter, onMouseLeave functions when mouse is moving over container element', () => {
			renderPoint({ onMouseOver: mockFn, onMouseEnter: mockFn, onMouseLeave: mockFn })
				.simulate('mouseover')
				.simulate('mouseenter')
				.simulate('mouseleave');
			expect(mockFn).toHaveBeenCalledTimes(3);
		})
	});
});