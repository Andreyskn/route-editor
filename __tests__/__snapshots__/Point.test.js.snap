// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Point /> Rendering renders Clone state correctly 1`] = `
<div
  className="point clone"
  style={
    Object {
      "transform": "translate3d(0, 10px, 0)",
    }
  }
>
  <button
    className="point-btn"
  />
  <div
    className="point-overlay"
  >
    <span
      className="point-text"
    >
      Test text
    </span>
  </div>
</div>
`;

exports[`<Point /> Rendering renders Default state correctly 1`] = `
<div
  className="point"
>
  <button
    className="point-btn"
  />
  <div
    className="point-overlay"
  >
    <span
      className="point-text"
    >
      Test text
    </span>
  </div>
</div>
`;

exports[`<Point /> Rendering renders Dragged state correctly 1`] = `
<div
  className="point dragged"
>
  <button
    className="point-btn"
  />
  <div
    className="point-overlay"
  >
    <span
      className="point-text"
    >
      Test text
    </span>
  </div>
</div>
`;
